### DISCUSSION TITLE
MOMA predicts most genes to be essential

### DISCUSSION THREAD

**Discussion by wshao1** [opencobra/cobrapy#1355] [Q&A]
Hi, I was curious to try out MOMA as an alternative to FBA for identifying essential genes. I applied `cobra.flux_analysis.single_gene_deletion` to Human-GEM, and the result is that almost all genes are identified as essential (2807/2897). The same holds true for contextualized human models I have generated. Is this expected behavior for MOMA?

Many of the post-gene KO growth-rates are zero, indicating that MOMA opted for zero growth rate compared to a suboptimal one. I reasoned that this might be because compensatory pathways are much longer and thus require activation of more fluxes, making it more "ideal" from MOMA's standpoint to simply have zero flux through the system. But how can it be that almost all gene KOs lead to zero growth rate?

**Comment by cdiener** [comment]
What medium/import bounds did you use here? I would imagine that this makes a difference. I just ran Recon3D and on completely open bounds there is only a single essential gene with MOMA, but the growth rates are also way too high. Another issue could be sink or demand reactions.

**Comment by wshao1** [comment]
I did not alter any bounds for the generic Human-GEM model, all exchange reactions are completely unbounded. For the contextualized models, the exchange reactions are also unbounded. Neither have demand/sink reactions, upon inspecting `model.sinks` or `model.demands`.

Can you explain why sink/demand reactions might pose an issue?

**Reply by cdiener** [reply]
Odd, seems to be specific to the model then, given the differences with Recon3. Which model are you using?

**Reply by wshao1** [reply]
I am using the most recent release of [Human-GEM (v1.16)](https://github.com/SysBioChalmers/Human-GEM/releases/tag/v1.16.0).

**Reply by cdiener** [reply]
Sinks can create situations where the model grows without any metabolite imports, masking essentiality where there should be one. Demands can create situations where low growth rates aren't possible because the demand lower bound can't be fulfilled, making some knockouts erroneously essential even though they should lead to a lower growth rate only.

**Reply by wshao1** [reply]
OK I just downloaded [Recon3D_3.01](https://www.vmh.life/#downloadview) and tried to replicate your MOMA essentiality results. I get 287 gene KOs resulting in zero growth. However, using FBA results in only a single essential gene. Here is my exact code, in case I am doing something differently.

```
recon3d = load_matlab_model('../raw_data/Recon3DModel_301.mat')
deletions_moma_recon3d = cobra.flux_analysis.deletion.single_gene_deletion(recon3d, method="moma")
num_essential_genes = deletions_moma_recon3d.loc[deletions_moma_recon3d['growth'] == 0.0].shape[0]
```

**Reply by cdiener** [reply]
I used `load_model("Recon3D")`. I could imagine the bounds are set differently in the mat file. What do you get for `recon3d.optimize()`?

**Reply by cdiener** [reply]
And what solver are you using?

**Reply by wshao1** [reply]
I get `Optimal solution with objective value 755.003`. I am using the Gurobi solver.

**Reply by wshao1** [reply]
I tried `load_model("Recon3D")` and got the same solution.

**Reply by cdiener** [reply]
Hmm, I don't get that result with CPLEX. That is odd.

**Reply by cdiener** [reply]
Oh do you mean you get the same optimum (755.xxx), or you still get 287 essential genes?

**Reply by cdiener** [reply]
For the Human-GEM I also get mostly non-essential ones. I did not run all because it takes long.

```
In [1]: from cobra.io import read_sbml_model

In [2]: hum = read_sbml_model("Human-GEM.xml")

In [3]: hum.solver
Out[3]: <optlang.cplex_interface.Model at 0x7f383992fed0>

In [4]: from cobra.flux_analysis import single_gene_deletion

In [5]: %time res = single_gene_deletion(hum, hum.genes[0::100], method="moma", processes=6)

CPU times: user 59.4 s, sys: 237 ms, total: 59.6 s
Wall time: 9min 2s

In [6]: 

In [6]: res
Out[6]: 
                  ids        growth     status
0   {ENSG00000167720}  1.384666e+02    optimal
1   {ENSG00000136250}  1.384666e+02    optimal
2   {ENSG00000113356}  1.384666e+02    optimal
3   {ENSG00000122729}  1.384666e+02    optimal
4   {ENSG00000079739}  1.384666e+02    optimal
5   {ENSG00000000419}  1.384666e+02    optimal
6   {ENSG00000128708}  1.384666e+02    optimal
7   {ENSG00000139209}  1.384666e+02    optimal
8   {ENSG00000109667}  1.384666e+02    optimal
9   {ENSG00000196305}  1.723836e-15    optimal
10  {ENSG00000165092}  1.384666e+02    optimal
11  {ENSG00000232264}  1.384666e+02    optimal
12  {ENSG00000105258} -0.000000e+00    optimal
13  {ENSG00000179115} -0.000000e+00    optimal
14  {ENSG00000170899}  1.384666e+02    optimal
15  {ENSG00000025800}  1.384666e+02    optimal
16  {ENSG00000184432}  1.384666e+02    optimal
17  {ENSG00000143149}  1.004651e+02  unbounded
18  {ENSG00000152556}  1.384666e+02    optimal
19  {ENSG00000067225}  1.384666e+02    optimal
20  {ENSG00000147160}  1.384666e+02    optimal
21  {ENSG00000132664} -0.000000e+00    optimal
22  {ENSG00000162688}  1.384666e+02    optimal
23  {ENSG00000101577}  1.384666e+02    optimal
24  {ENSG00000173915}  1.181962e+02    optimal
25  {ENSG00000198805}  1.047539e+02    optimal
26  {ENSG00000117054}  1.384666e+02    optimal
27  {ENSG00000094841}  1.384666e+02    optimal
28  {ENSG00000158296}  1.384666e+02    optimal 
```

**Reply by wshao1** [reply]
Sorry I meant same optimum. I will give CPLEX a try to see if that changes anything!

**Reply by cdiener** [reply]
Hmmm, it shouldn't though because MOMA calculate the reference solution using pFBA so it should be comparable across solvers.

**Reply by wshao1** [reply]
OK it took a long time to run, but CPLEX does give a different result: 80 essential genes for Human-GEM. I was also able to replicate your result of 1 essential gene for Recon3D.

How concerning is the discrepancy between the two solvers?

**Reply by cdiener** [reply]
Yeah somewhat. MOMA depends a lot on the reference solution. So the first thing I would check is whether the pFBA solution is the same with both solvers (since pFBA is not guaranteed to be unique). I don't have a Gurobi license so I can't really debug it though. If the reference solution is indeed the same, than one of the two solver interfaces is wrong, because the quadratic problem in MOMA has a unique solution. And that would be something we would need to fix for sure.

**Reply by wshao1** [reply]
You were right, the pFBA solutions are indeed different (about half of the fluxes differ between the two solvers). Thanks for pointing out the sensitivity of MOMA. Interpretation-wise, is this just an inherent limitation of an under-constrained problem? If both pFBA solutions generated by the two solvers are equally valid, and MOMA is working as designed, then I don't have any reason on theoretical grounds to choose one solver over the other correct?

**Reply by cdiener** [reply]
> You were right, the pFBA solutions are indeed different (about half of the fluxes differ between the two solvers). Thanks for pointing out the sensitivity of MOMA. 

Phew, that is a bit less concerning :sweat_smile:  

> Interpretation-wise, is this just an inherent limitation of an under-constrained problem? If both pFBA solutions generated by the two solvers are equally valid, and MOMA is working as designed, then I don't have any reason on theoretical grounds to choose one solver over the other correct?

Yeah kind of. MOMA just minimizes the distance to a reference solution, but does not help to pick a good reference solution. Instead of choosing a solver, it would be more about choosing a reference flux distribution/solution here. For instance, does one seem more correct than the other? Maybe one produces a metabolite one would expect and the other doesn't. Also, the constraints are probably an issue as well. pFBA might give you a more unique solution when you apply realistic import bounds (like a medium).

**Reply by wshao1** [reply]
OK that makes sense. Thank you so much for explaining! 😊
